### PR TITLE
Migrate project activity to LFNext

### LINKED ISSUE #1393: Migrate `app/activity` to LFNext
discussion: #1394

### PR DESCRIPTION
**Pull request by longrunningprocess**
## Description

This PR seeks to migrate the project activity capability to LFNext.  While at it, discussion arose about a "Project landing page".  That is also included in this work.  Refer to #1394 for further discussion

Fixes #1393 

### Type of Change

Only keep lines below that describe this change, then delete the rest.

- New feature (non-breaking change which adds functionality)
- UI change

## Screenshots

This video runs through the look of it, mobile views and all of the links back into this page.

https://user-images.githubusercontent.com/4412848/171720727-00b36ca5-b32a-4598-9b54-363a0bd9b540.mov

forgot to include some of the h-scroll behavior of the activity table on mobile, that demo is here:

https://user-images.githubusercontent.com/4412848/171729717-8faec45a-7064-40fc-8b1d-15122efb5c15.mov

## Testing on your branch

- [ ] Start a new project without any activity and ensure the activity table reflects that.
- [ ] With your new project, add a word and some activity to ensure the most recent data shows up on the activity table.
- [ ] Pick a project with known activity and ensure only the last 30 days is initially shown
- [ ] When viewing a project with activity, ensure the data is sorted by date, then by user within the same date
- [ ] When viewing a project with activity, ensure clicking "show all" retrieves all activity
- [ ] Ensure the project link under the "My Projects" dropdown in the header takes you to the correct project landing page
- [ ] Ensure while on the project landing page, when you click "work on this project" you are taken to the editor view for the correct project
- [ ] Ensure the project links from under the "My Projects" page (`/app/projects`) takes you to the correct project landing page
- [ ] Ensure the project link in the breadcrumb takes you to the correct project landing page
- [ ] Ensure you cannot see the activity for a project when you are not logged in
- [ ] Ensure you cannot see the activity for a project you are not allowed to view
- [ ] Create a link for sharing the project and ensure when the link is used it redirects to the correct project's landing page
- [ ] Create a project via upload and check the activity pages for it

## Checklist

- [x] I have performed a self-review of my own code
- [x] I have reviewed the title/description of this PR which will be used as the squashed PR commit message
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added tests that prove my fix is effective or that my feature works

## qa.languageforge.org testing

Reviewers: add/replace your name below and check the box to sign-off/attest the feature works as expected on qa.languageforge.org

- [ ] Reviewer1 (YYYY-MM-DD HH:MM)
- [ ] Reviewer2 (YYYY-MM-DD HH:MM)

### CODE REVIEW

**Comment by rmunn** [comment]
Just built this branch locally and ran it, and it didn't find activity from 5 months ago on one of my test projects. Here's what happens when I clicked the "Activity Feed" button on one entry from my test-rmunn-05 project:

image

Note the date: January 3, 2022, quite a bit more than 30 days ago. If I understand the "When viewing a project with activity, ensure clicking "show all" retrieves all activity" bullet point correctly, that should have produced a "See all" button, but it wasn't visible:

image

(I trimmed that screenshot horizontally so as to fit into the GitHub comments UI, but there were no buttons other than the "Work on this project" button).

@longrunningprocess, I'll send you the raw Mongo data for this project if you want to use it for testing.

File: next-app/src/lib/forms/Button.svelte
Comment (rmunn) [inline next-app/src/lib/forms/Button.svelte:6]: https://svelte.dev/docs#template-syntax-attributes-and-props recommends against using `$$props` for perfornance's sake (emphasis mine):

> `$$props` references all props that are passed to a component, including ones that are not declared with export. **It is not generally recommended**, as it is difficult for Svelte to optimise. But it can be useful in rare cases – for example, when you don't know at compile time what props might be passed to a component.

Is there any reason not to do `export let class: string = ''` in the component and then do `{ class }` here?

**Review by rmunn — COMMENTED** [review]
Haven't yet looked closely at the Svelte / Svelte-Kit stuff — that's next — but I have some comments on the rest of the code. For the Svelte-Kit stuff, I plan to look at the code in context inside VS Code rather than reading only the changes in the GitHub review UI, as I think I'll understand it better that way and therefore be able to review this more fully. But there's no reason to make you wait for my initial reactions.

**Comment by longrunningprocess** [comment]
> Just built this branch locally and ran it, and it didn't find activity from 5 months ago on one of my test projects. Here's what happens when I clicked the "Activity Feed" button on one entry from my test-rmunn-05 project:
> 
> image
> 
> Note the date: January 3, 2022, quite a bit more than 30 days ago. If I understand the "When viewing a project with activity, ensure clicking "show all" retrieves all activity" bullet point correctly, that should have produced a "See all" button, but it wasn't visible:
> 
> image
> 
> (I trimmed that screenshot horizontally so as to fit into the GitHub comments UI, but there were no buttons other than the "Work on this project" button).
> 
> @longrunningprocess, I'll send you the raw Mongo data for this project if you want to use it for testing.

yes, that is a great find, I did not think that condition through properly.  I'll put in a fix for it.

File: next-app/src/lib/forms/Button.svelte
Comment (longrunningprocess) [reply next-app/src/lib/forms/Button.svelte]: I am aware of their advice on the use of `$$props` but have never found it to be an issue but am happy to incorporate a technique I've used in the past for this sort of thing.  Take a look and let me know what you think (https://github.com/sillsdev/web-languageforge/pull/1416/commits/18140fc6e8199d0a402c4be9aaa8da5b0390db2f).  Since `class` is a reserved word, we have to do some yucky workaround which is why I typically just stick with `$$props`.

File: src/Api/Service/Sf.php
Comment (longrunningprocess) [reply src/Api/Service/Sf.php:420]: this and the following two comments also have me re-thinking that "activity feed" button altogether... @megahirt do you think we should leave that button in the legacy app or hide it and just leave all these methods Robin is suggesting we not remove?

Comment (megahirt) [reply src/Api/Service/Sf.php:420]: This PR is actually about replacing the legacy activity feed with a new activity feed provided by LFNext.

1) The new activity feed is for the current project only.  A "project dashboard" of sorts.
2) The old activity feed can be removed from the legacy app.  No need to keep it around.
3) The PHP function valid_activity_types should be removed since it is not used in the new LFNext activity feed.

A few things currently missing from this PR:

1) clean up angularJS files no longer needed: angularJS services and templates related to activity feed
2) remove references to validActivityTypes

Comment (longrunningprocess) [reply src/Api/Service/Sf.php:420]: @megahirt @rmunn please let me know you've done this so I can resolve this conversation.

File: src/Api/Model/Shared/Dto/ActivityListDto.php
Comment (longrunningprocess) [reply src/Api/Model/Shared/Dto/ActivityListDto.php:40]: this might be ok depending upon the conversation in Sf.php, resolving this one for now.

File: src/Api/Model/Shared/ActivityModel.php
Comment (longrunningprocess) [reply src/Api/Model/Shared/ActivityModel.php:127]: this might be ok depending upon the conversation in Sf.php, resolving this one for now.

File: docker/next-app/Dockerfile.next-app
Comment (longrunningprocess) [reply docker/next-app/Dockerfile.next-app]: not waiting on you to resolve this since I took your suggestion.

File: next-app/src/lib/forms/Button.svelte
Comment (longrunningprocess) [reply next-app/src/lib/forms/Button.svelte]: mind reviewing my last reply @rmunn ?

File: docker/next-app/dev/Dockerfile
Comment (longrunningprocess) [reply docker/next-app/dev/Dockerfile:5]: I'll go ahead an resolve this one for you since it's what you were suggesting.

**Comment by josephmyers** [comment]
After starting in on the review, I pretty quickly realized I wasn't going to be able to contribute much on the code side. But in testing, I have a couple comments about the UI (since that's what I'm good at), and you can feel free to disregard them or split them out as out-of-scope:

- What's the recommended way to reach the Activity page? In my usage, given navigation tree "My Projects / Project / Edit", I wanted "Project" to take me to some sort of home page or Activity page, not the page I was already on.
- At what point do the entries cut off and make the Show All button useful? I made ~50 entries and was still able to see all w/o clicking the button
- Additionally, (you've probably thought of this) but I'm sure the list of entries will become quite overwhelming. is there some intermediate stage between showing some and showing all, e.g. pages?
- Would it be helpful to linkify the entries/users?
- It's weird to me that the Activity page uses a different, non-interactive "logo" for Language Forge at the top-left. is this part WIP?
- General design: what do you think about altering slightly the look of the clickable/interactive items to indicate to the user they're different from simple text? as someone unfamiliar with the program, i found myself clicking around a lot on random things. this is especially important on pointer-less mobile platforms

**Comment by longrunningprocess** [comment]
> After starting in on the review, I pretty quickly realized I wasn't going to be able to contribute much on the code side. But in testing, I have a couple comments about the UI (since that's what I'm good at), and you can feel free to disregard them or split them out as out-of-scope:

Excellent, thank you!


> 
> * What's the recommended way to reach the Activity page? In my usage, given navigation tree "My Projects / Project / Edit", I wanted "Project" to take me to some sort of home page or Activity page, not the page I was already on.

via the breadcrumbs (the project name link in "My Projects / <project name> / Edit" will take you to the project "landing" page which includes the activity...the hope is that page will evolve to include dashboard type info about the project)... it occurs to me as I'm writing this that maybe you tested this in staging already where the old code is still running... the code in this branch won't be in staging until the PR is approved.  I initially thought maybe you were testing this locally).


> * At what point do the entries cut off and make the Show All button useful? I made ~50 entries and was still able to see all w/o clicking the button
> * Additionally, (you've probably thought of this) but I'm sure the list of entries will become quite overwhelming. is there some intermediate stage between showing some and showing all, e.g. pages?

The default activity is for the last 30 days... there is no intelligence to this for now, e.g., give me the last 30 days but no more than 20 rows of activity.  That would be nice if it's an issue for projects, I'm not sure what to expect from users yet sinceI'm not aware of this activity even being used very much today.

> * Would it be helpful to linkify the entries/users?

Possibly at a future date.  I think I'm inclined to get the first pass of this page in front of users to start getting some of their feedback before we begin building links away.  Right now the basic use case is to start on the project landing page -> begin working on it.  I'd like to keep talking about other use cases though, thanks for thinking about that.

> * It's weird to me that the Activity page uses a different, non-interactive "logo" for Language Forge at the top-left. is this part WIP?

I find the interactivity of that link a little strange as it is in the legacy app, i.e., when I click it I would expect to go to the home page but it actually puts me in the editor... 🤷🏻  I'd like to talk about it more though.  The simpler approach to me is just to leave it as informative/orienting rather than functional.  Functional things, like linking away, are currently found in the right-hand menu.

> * General design: what do you think about altering slightly the look of the clickable/interactive items to indicate to the user they're different from simple text? as someone unfamiliar with the program, i found myself clicking around a lot on random things. this is especially important on pointer-less mobile platforms

I hear that, let's discuss the design system around those more.  I'm going with the defaults at this point but things like that can be changed at an app-wide level:  https://daisyui.com/components/link and/or https://daisyui.com/components/button

File: next-app/src/lib/forms/Button.svelte
Comment (rmunn) [reply next-app/src/lib/forms/Button.svelte]: Yes, the `export { className as class }` (or `classes` here, doesn't really matter what it's called) technique is nice, and I see that's what other people are recommending for this situation. I like it, and I feel okay with resolving this conversation. Sorry it lingered so long.

File: src/Api/Service/Sf.php
Comment (rmunn) [reply src/Api/Service/Sf.php:420]: If we're removing the "Activity Feed" button entirely from the legacy app, then there is more code that can be removed:

* `activity_list_dto_for_lexical_entry` in Sf.php and RightsHelper.php (was used in the `listActivityForLexicalEntry` function in activity.service.ts, but activity.service.ts is entirely gone)
* `getActivityForOneLexEntry` and `getActivityForLexEntry` methods in ActivityListDto.php (only used by the `activity_list_dto_for_lexical_entry` code in Sf.php)
* `ActivityListModelByLexEntry` class in ActivityListDto.php (only used in the `getActivityForLexEntry` method which is being removed)

I actually like having a per-entry activity feed, so I don't quite agree with removing that button or the code that feeds it, since it might be a little while before the editor view gets converted to LFNext, and until that point, it's still useful to have activity-list code in the legacy app. So I agree with @longrunningprocess 's initial design of keeping the legacy-app "Activity Feed" button around. Let's have that conversation next time we're able to all meet together, and decide whether the activity feed should really be ripped out of the legacy app or not.

**Comment by rmunn** [comment]
Was able to get back to testing this again, and here are a few other issues that I found:

activity-feed-2022-07-06

* Date column isn't sorted (one 2020 item appeared at the top, above the 2022 items). Apparently we don't have a sort order in the Mongo query, and instead do the sorting afterwards in the `sortActivity` method of ActivityListDto.php. That method has been [removed in the current version of the PR](https://github.com/sillsdev/web-languageforge/pull/1416/files#diff-c88de02ee63d32b6eb10744092d53b2e5233be67efaaee79d13d07b70722d77dL130), and should probably be put back, along with the `uasort` calls that reference it.
* Two activity types (`add_lex_reply` and `add_lex_comment`) don't have a human-readable version yet, so the programmatic IDs leaked through into the UI.

Thanks for all your work so far, and sorry that I took nearly three weeks to look at this again!

**Comment by rmunn** [comment]
@longrunningprocess See #1430 for a few fixes for some TypeScript errors. There were just enough changes that I didn't really want to put this as a code suggestion in GitHub as that would have involved a bunch of files. I created it as a PR onto your branch, rather than just pushing a commit onto your branch, since I know you rebase often and I didn't want to mess up your rebase workflow by having a commit of mine on your branch that you weren't expecting.

**Comment by longrunningprocess** [comment]
> * Two activity types (`add_lex_reply` and `add_lex_comment`) don't have a human-readable version yet, so the programmatic IDs leaked through into the UI.

thank you, I corrected these in [62f498c](https://github.com/sillsdev/web-languageforge/pull/1416/commits/62f498ce34c6a77e3fc5e25ffc6dcbe5cde2883a).  My plan was to simply updates those as they appear since I can't really tell which of them in `src/Api/Model/Shared/ActivityModel.php` are actually used still or relevant.

**Comment by rmunn** [comment]
> I'm actually doing the sorting client side (https://github.com/sillsdev/web-languageforge/blob/next/project-activity/next-app/src/routes/projects/%5Bproject_code%5D/_components/Activity.svelte#L29) and I thought using epoch was a safe approach, maybe not... I thought I looked into the multi-level sort in mongo and didn't go with it for some reason I can't remember at this point.

Figured it out during our standup meeting: it's sorting by the locale date string, meaning that the string "6/27/2020" was sorting as "larger" than the string "3/1/2022" because 6 > 3. I agree with the idea of sorting by date first, then user, then time, but we should make sure the date being sorted on is a Javascript date and not a locale-formatted date string. (Unless the locale is YYYY-MM-DD, which would have sorted correctly :-) ).

**Comment by longrunningprocess** [comment]
> 1b44408



> > I'm actually doing the sorting client side (https://github.com/sillsdev/web-languageforge/blob/next/project-activity/next-app/src/routes/projects/%5Bproject_code%5D/_components/Activity.svelte#L29) and I thought using epoch was a safe approach, maybe not... I thought I looked into the multi-level sort in mongo and didn't go with it for some reason I can't remember at this point.
> 
> Figured it out during our standup meeting: it's sorting by the locale date string, meaning that the string "6/27/2020" was sorting as "larger" than the string "3/1/2022" because 6 > 3. I agree with the idea of sorting by date first, then user, then time, but we should make sure the date being sorted on is a Javascript date and not a locale-formatted date string. (Unless the locale is YYYY-MM-DD, which would have sorted correctly :-) ).

I believe I've addressed this in [1b44408](https://github.com/sillsdev/web-languageforge/pull/1416/commits/1b44408a431362a2211585bb7060c4836c118123)[1b44408](https://github.com/sillsdev/web-languageforge/pull/1416/commits/1b44408a431362a2211585bb7060c4836c118123), please let me know what you think @rmunn .

File: src/Api/Service/Sf.php
Comment (longrunningprocess) [reply src/Api/Service/Sf.php]: I reverted that stuff today @rmunn , things should be back the way they were now.

File: next-app/src/lib/forms/Button.svelte
Comment (megahirt) [reply next-app/src/lib/forms/Button.svelte:6]: see my other comment above in a different line.  I am fine with the resolution of keeping $$props.class, just recommending an explanatory comment to that effect.

File: next-app/src/lib/icons/ImagesIcon.svelte
Comment (megahirt) [inline next-app/src/lib/icons/ImagesIcon.svelte:1]: I'm commenting here because I don't know how to ask GitHub reviews to show me the commit where this file was introduced (I'm not on a dev machine so don't have access to the repo locally).  I'm guessing the commit message would say where these icons came from and why they were committed.

So that's essentially my question: are these icons assets that we should commit into our repo or should they be pulled in as some sort of dependency?  Sorry if this is a dumb question.  I was hoping to figure this out on my own by reading commit messages in GitHub but couldn't see a good way to do that.

File: next-app/src/routes/__layout.svelte
Comment (megahirt) [inline next-app/src/routes/__layout.svelte:23]: I've noticed that some files, like this one, contain just whitespace changes.  The changes are great, and I assume that you have just run prettier or some other code formatter on the file.  I lament that we don't yet have a consistent code formatting solution that all devs are using.  I know it's not hard - just finding it hard to settle on Prettier, get all our code processed, and then enforce it at the commit level.

File: next-app/src/routes/projects/[project_code]/index.ts
Comment (megahirt) [inline next-app/src/routes/projects/[project_code]/index.ts:4]: I was wondering what the significance of capital GET and PUT is, compared to their lowercase equivalents?  Is this following a convention?

**Review by megahirt — COMMENTED** [review]
This is an incomplete code-review.  I've read through about half of it and made some comments.  I've run out of time after the kids are in bed...so will pick this up again tomorrow.  Great job so far @longrunningprocess !

File: next-app/src/lib/PageHeader.svelte
Comment (longrunningprocess) [reply next-app/src/lib/PageHeader.svelte:1]: Instead of documenting this everywhere it occurs, what would think of this approach:  [0956b5c](https://github.com/sillsdev/web-languageforge/pull/1416/commits/0956b5ca30740378b9bf9d10e3d8605037c18454)

File: next-app/src/lib/error/Error.svelte
Comment (longrunningprocess) [reply next-app/src/lib/error/Error.svelte:5]: I'm looking at this now actually.

File: next-app/src/lib/icons/ImagesIcon.svelte
Comment (longrunningprocess) [reply next-app/src/lib/icons/ImagesIcon.svelte:1]: I suppose it would be a stylistic question.  If we made the SVG a separate file and simply imported it into this component it would only result in more to manage.  These SVG's can be custom or come from some free source like daisyUI or material icons... if you'd like to document the source on these, that is certainly something we can do in these files as a comment.

File: next-app/src/routes/__layout.svelte
Comment (longrunningprocess) [reply next-app/src/routes/__layout.svelte:23]: sure, we can incorporate something like that if you'd like.  If it bothers you on code reviews, simply use the option available in GH to ignore whitespace changes during the review.  I can show you where to find that if you don't see it.

File: next-app/src/routes/projects/[project_code]/index.ts
Comment (longrunningprocess) [reply next-app/src/routes/projects/[project_code]/index.ts]: that is a recent breaking change to SvelteKit.  They've decided to following the HTTP convention on these.  You'll notice my data wrappers were already doing this just because I personally like the matchup with HTTP and I wanted to be able to use `DELETE` in its full form rather than `del`.  Here's a broader conversation on it for your reference:  https://github.com/sveltejs/kit/discussions/5359

File: src/Api/Model/Shared/Dto/ActivityListDto.php
Comment (megahirt) [inline src/Api/Model/Shared/Dto/ActivityListDto.php:54]: Note to self: this will likely break the existing Activity feed that we are replacing.  So doing a "fast revert" using caddy routing probably won't work since we have changed the data structure the server is responding with.

File: src/Api/Service/Sf.php
Comment (megahirt) [inline src/Api/Service/Sf.php:549]: I guess I'm OK with this for the moment, but this is horribly inefficient to call the lexdbedto(encode) for every page load.  The method was designed to send all dictionary data to the client to be stored locally.  We should talk through our offline use case and also begin to develop a strategy where we make use of the existing indexeddb data stored locally on the client already.

I see this addition as a best effort by you to get the data you were asked to display in the dashboard while going the SSR route and I was essentially unavailable for consult or development help.

**Review by megahirt — COMMENTED** [review]
I have finished going through the code...Now I just need to run it up on my machine :)

File: next-app/src/routes/projects/[project_code]/index.ts
Comment (megahirt) [reply next-app/src/routes/projects/[project_code]/index.ts]: ok thanks for the info!

File: next-app/src/routes/__layout.svelte
Comment (megahirt) [reply next-app/src/routes/__layout.svelte:23]: yes, that's true about the whitespace changes.  I can use that filter.

I definitely want us to all be using a code formatter.

File: next-app/src/lib/icons/ImagesIcon.svelte
Comment (megahirt) [reply next-app/src/lib/icons/ImagesIcon.svelte:1]: Yes ideally we document in the file itself where the code came from, if we didn't write it ourselves (attribution)

File: next-app/src/lib/PageHeader.svelte
Comment (megahirt) [reply next-app/src/lib/PageHeader.svelte:1]: yes, this is fine.  Thank you!

File: src/Api/Model/Shared/Dto/ActivityListDto.php
Comment (longrunningprocess) [reply src/Api/Model/Shared/Dto/ActivityListDto.php:54]: yeah, if I'm understanding you, you're right we couldn't just tweak the proxy but if we were in a situation where we need to rollback, I'd think we'd just go back a version in the images which would not have this.

File: src/Api/Service/Sf.php
Comment (longrunningprocess) [reply src/Api/Service/Sf.php:549]: ok, sounds like we need to talk about that more.  Putting a strategy in place to replicate the db locally and keep deltas in sync across both sources is quite an undertaking...

**Review by josephmyers — APPROVED** [review]
I forgot to actually approve this after we talked in July. From the UI side of things, this is good to go

File: src/Api/Service/Sf.php
Comment (megahirt) [reply src/Api/Service/Sf.php:549]: yeah, we need to talk more about strategy, but this won't block the PR from being merged.

**Review by megahirt — APPROVED** [review]
Good to go @longrunningprocess !

We should check on package-lock.json after this merges to see if we need to re-create it.  I don't have full confidence in git merge to just assume it will be fine.

**Comment by longrunningprocess** [comment]
> Good to go @longrunningprocess !
> 
> We should check on package-lock.json after this merges to see if we need to re-create it. I don't have full confidence in git merge to just assume it will be fine.

I've merged it and all tests seem to be good, maybe once we test it a bit in staging we can have confidence in git's merge?

**Pull request merged**
